scipy/optimize/optimize.py

```diff
@@ -744,10 +744,13 @@ def _minimize_neldermead(func, x0, args=(), callback=None,
         sim = np.clip(sim, lower_bound, upper_bound)
 
     one2np1 = list(range(1, N + 1))
-    fsim = np.empty((N + 1,), float)
 
-    for k in range(N + 1):
-        fsim[k] = func(sim[k])
+    fsim = np.array([func(sim[k]) for k in range(N + 1)], dtype=float)
```

Comment:
Does this reformulation guarantee behaviour equivalent to what we had before? For example, if for some inputs the function `func` returns a `(1, 1)` array and for others it returns a `(1, )` array, this should fail. Not a optimize expert, but I wonder if we need to preserve that for backwards compatibility. If it doesn't incur much overhead, maybe add `.item()` in the for loop we had previously?

Reply:
> func returns a (1, 1) array and for others it returns a (1, ) array, this should fail.

You're right, it would fail now and wouldn't have failed before the PR. The workaround
```python
np.array([np.asarray(func(sim[k])).item() for k in range(N + 1)], dtype=float)
```
may look a bit extreme at first sight, but you're right, maybe we _should_ actually wrap every user function call in `np.asarray(f(x)).item()`. We can also apply that to the rest of the calls to save some code.